velox/common/file/FileSystems.h

```diff
@@ -0,0 +1,87 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string_view>
+
+// Private. Return the external filesystem name given a tag.
+#define _VELOX_REGISTER_FILESYSTEM(tag) registerFileSystem_##tag
+
+// Registers a filesystem associated with a given tag.
+#define VELOX_REGISTER_FILESYSTEM(tag)             \
+  {                                                \
+    extern void _VELOX_REGISTER_FILESYSTEM(tag)(); \
+    _VELOX_REGISTER_FILESYSTEM(tag)();             \
+  }
+
+// Declares a filesystem associated with a given tag.
+#define VELOX_DECLARE_FILESYSTEM(tag, schemaMatcher, filesystemGenerator) \
+  void _VELOX_REGISTER_FILESYSTEM(tag)() {                                \
+    registerFileSystem((schemaMatcher), (filesystemGenerator));           \
+  }
+
+namespace facebook::velox {
+class Config;
+class ReadFile;
+class WriteFile;
+} // namespace facebook::velox
+
+namespace facebook::velox::filesystems {
+
+// An abstract FileSystem
+class FileSystem {
+ public:
+  FileSystem(std::shared_ptr<const Config> config)
+      : config_(std::move(config)) {}
+  virtual ~FileSystem() {}
+  // Returns the name of the File System
+  virtual std::string name() const = 0;
+  // Returns a ReadFile handle for a given file path
+  virtual std::unique_ptr<ReadFile> openFileForRead(std::string_view path) = 0;
+  // Returns a WriteFile handle for a given file path
+  virtual std::unique_ptr<WriteFile> openFileForWrite(
+      std::string_view path) = 0;
+
+ protected:
+  std::shared_ptr<const Config> config_;
+};
+
+std::shared_ptr<FileSystem> getFileSystem(
+    std::string_view filename,
+    std::shared_ptr<const Config> config);
+
+// FileSystems must be registered explicitly.
+// The registration function takes two parameters:
+// a std::function<bool(std::string_view)> that says whether the registered
+// FileSystem subclass should be used for that filename,
+// and a lambda that generates the actual file system.
+void registerFileSystem(
+    std::function<bool(std::string_view)> schemeMatcher,
+    std::function<std::shared_ptr<FileSystem>(std::shared_ptr<const Config>)>
+        fileSystemGenerator);
+
+// Register the local filesystem.
+void registerLocalFileSystem();
+
+// Register all filesystems.
+// Each registered file system is tried in the
+// order it was registered, so keep this in mind if multiple file systems could
+// match the same filename.
+void registerAllFileSystems();
```

Comment:
What would be the reason for having this method? Do you expect to update it every time a new FileSystem is added ? Why can't we remove this?

Reply:
I felt this would be a nice convenient function for applications that want all the available FileSystems. Yes, I expect it to be updated every time a new file system is added.
Once an application includes this in its code, it automatically gets any new file system added later. Else it is a manual process to update the application.

Reply:
> I expect it to be updated every time a new file system is added.

Does that mean that this component will have dependencies on all file-system related components, e.g. S3, HDFS, etc.? Is this desirable?

Reply:
You have a fair point. We can avoid dependency by gating with directives.
```
#if defined VELOX_S3 
registerS3FileSystem();
#endif 
```
We will be adding those CMake options. I guess it is a trade-off here.

Reply:
@majetideepak Will it be better for each fileSystem to register by themself instead of calling `registerAllFileSystems` in one filesystem? 

For example, LocalFileSystem.registerAllFileSystems only register local file system now